packages/next/src/build/webpack-config.ts

```diff
@@ -1939,87 +1854,158 @@ export default async function getBaseWebpackConfig(
     },
     module: {
       rules: [
+        // This first rules is to set layers and conditions. Be careful if adding any conditions
+        // to any later rules as they might match both and you end up with more than anticipated.
+        // The reason this rule as to be a oneOf is boundary modules can match more than one of these
+        // rules (i.e. they set the layer to client but they are part of the server layer and end up
+        // with both sets of conditions since they merge via "..." extension)
+        // Additionally please not that we intentionally do not use alias configuraiton here. We have
+        // Two plugins NextResolveAppAliasPlugin and NextResolveVendorPlugin which handle aliasing
+        // and vendoring for all App Router builds.
         ...(hasAppDir
           ? [
               {
-                // Make sure that AsyncLocalStorage module instance is shared between server and client
-                // layers.
-                layer: WEBPACK_LAYERS.shared,
-                test: asyncStoragesRegex,
-              },
-              // Convert metadata routes to separate layer
-              {
-                resourceQuery: new RegExp(
-                  WEBPACK_RESOURCE_QUERIES.metadataRoute
-                ),
-                layer: WEBPACK_LAYERS.metadataRoute,
-              },
-              {
-                // Ensure that the app page module is in the client layers, this
-                // enables React to work correctly for RSC.
-                layer: WEBPACK_LAYERS.client,
-                test: /next[\\/]dist[\\/](esm[\\/])?server[\\/]future[\\/]route-modules[\\/]app-page[\\/]module/,
-              },
-              {
-                // All app dir layers need to use this configured resolution logic
-                issuerLayer: {
-                  or: [
-                    WEBPACK_LAYERS.server,
-                    WEBPACK_LAYERS.client,
-                    WEBPACK_LAYERS.appClient,
-                    WEBPACK_LAYERS.action,
-                    WEBPACK_LAYERS.shared,
-                  ],
-                },
-                resolve: {
-                  alias: {
-                    // Alias next/head component to noop for RSC
-                    [require.resolve('next/head')]: require.resolve(
-                      'next/dist/client/components/noop-head'
-                    ),
-                    // Alias next/dynamic
-                    [require.resolve('next/dynamic')]: require.resolve(
-                      'next/dist/shared/lib/app-dynamic'
+                oneOf: [
```

Comment:
Potentially related to CSS issue

Reply:
Hmm the oneOf behavior is odd. I contributed a webpack config that wrapped a single rule in oneOf which should be functionally equivalent to not wrapping it and the css fails to load. leaving it unrwapped it is bundled just fine. I can't make sense of it. But regardless I have a config that works without the oneOf now so I think we are potentially good to move forward